### PR TITLE
[WIP] Frame + animate API

### PR DESCRIPTION
**Pull request by rreusser**
Progress:

- [x] SVG element persistence
- [x] frame API
 - [x] `Plotly.addFrames`
 - [x] `Plotly.deleteFrames`
- [ ] frame diff merging logic
- [ ] `Plotly.transition`
- [ ] `Plotly.animate`

### CODE REVIEW

**Comment by rreusser** [comment]
@etpinard @mdtusz I've scaffolded out [some top-level keyframe management functions](https://github.com/rreusser/plotly.js/blob/77127d7862a465047a905a1e89c757aa85545c4a/src/plot_api/plot_api.js#L2587-L2633
) with the goal of keeping everything nice and undoable.

To set up keyframes:

1. If keyframes are present, a slider or similar control optionally shows up?
2. Create a keyframe using the UI (`createKeyframe()`)
3. Modify the graph (changes in the UI are mirrored to a globally selected keyframe via `modifyKeyframe()`?)

To animate to a keyframe:

1. Compute the state to which to animate (implement a function (`computeKeyframe()`?) that merges a keyframe with its base keyframe)
2. use `.animate()` to apply the changes

**Comment by rreusser** [comment]
Added tests and minimal implementation of keyframe modification logic. It needs better merging of new properties, but it at least obeys basic setting/undo/redo semantics:

https://github.com/rreusser/plotly.js/blob/animate-api-take-2/test/jasmine/tests/keyframe_api_test.js

File: test/jasmine/tests/keyframe_api_test.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+function failBecause(message) {
+    return function() {
+        expect(message).toBe(true);
+        return Promise.resolve();
+    };
+}
+
+describe('Test keyframe api', function() {
+    'ust strict';
```
Comment (timelyportfolio) [inline test/jasmine/tests/keyframe_api_test.js]: `use_strict` ?

Comment (rreusser) [inline test/jasmine/tests/keyframe_api_test.js]: Could elevate to the file level. Just makes messy thighs actually invalid.

Comment (rreusser) [inline test/jasmine/tests/keyframe_api_test.js]: Ahhh. Good catch.

**Comment by rreusser** [comment]
@etpinard I've been force-pushing over the same commit to keep the history from getting out of control, but if you're going to be working off of this branch, just let me know and I'll switch to more verbose commit history.

I'm currently modifying what I've done so far and scaffolding out the pieces we talked about. My first project is getting the frame API working as described. (merging etc.)

If you have time to contribute, the part I probably need the most help with is assimilating new data into the state of the plot. In other words, taking a data array and transferring x, y, colors, etc. into calcdata. I took some shortcuts and liberties that worked fine but probably need to be cleaner in the real thing, and it gets to the heart of nontrivial memory management that perhaps needs an experienced pair of eyes to fully appreciate all the caveats.

Specifically, here's the data transfer part I really didn't like on the first go. Felt very messy and ad-hoc: https://github.com/plotly/plotly.js/pull/550/files#diff-2941ab69a12080c0633ff4ac8ea3aa83R1577

**Pull request closed without merging**